lib/pg_eventstore/queries.rb

```diff
@@ -0,0 +1,140 @@
+    # @param stream [PgEventstore::Stream]
+    # @return [PgEventstore::Stream, nil] persisted stream
+    def find_stream(stream)
+      find_sql = <<~SQL
+        SELECT * FROM streams WHERE streams.context = $1 AND streams.stream_name = $2 AND streams.stream_id = $3
+          LIMIT 1
+      SQL
+      pgresult = connection.with do |conn|
+        conn.exec_params(find_sql, stream.to_a)
+      end
+      PgEventstore::Stream.new(**pgresult.to_a.first.transform_keys(&:to_sym)) if pgresult.ntuples == 1
+    end
+
+    # @param stream [PgEventstore::Stream]
+    # @return [PgEventstore::RawStream] persisted stream
+    def create_stream(stream)
+      create_sql = <<~SQL
+        INSERT INTO streams (context, stream_name, stream_id) VALUES ($1, $2, $3) RETURNING *
+      SQL
+      pgresult = connection.with do |conn|
+        conn.exec_params(create_sql, stream.to_a)
+      end
+      PgEventstore::Stream.new(**pgresult.to_a.first.transform_keys(&:to_sym))
+    end
+
+    # @return [PgEventstore::Stream] persisted stream
+    def find_or_create_stream(stream)
+      find_stream(stream) || create_stream(stream)
+    end
+
+    # @see PgEventstore::Client#read for more info
+    # @param stream [PgEventstore::Stream]
+    # @param options [Hash]
+    # @return [Array<PgEventstore::Event>]
+    def stream_events(stream, options)
+      exec_params = events_filtering_builder(stream, options).to_exec_params
+      pgresult = connection.with do |conn|
+        conn.exec_params(*exec_params)
+      end
+      deserializer.deserialize_many(pgresult)
+    end
+
+    # @param stream [PgEventstore::Stream] persisted stream
+    # @param event [PgEventstore::Event]
+    # @return [PgEventstore::Event]
+    def insert(stream, event)
+      serializer.serialize(event)
+
+      attributes = event.options_hash.slice(:id, :type, :data, :metadata, :stream_revision, :link_id).compact
+      attributes[:stream_id] = stream.id
+
+      sql = <<~SQL
+        INSERT INTO events (#{attributes.keys.join(', ')}) 
+          VALUES (#{(1..attributes.values.size).map { |n| "$#{n}" }.join(', ')}) 
+          RETURNING *
+      SQL
+
+      pgresult = connection.with do |conn|
+        conn.exec_params(sql, attributes.values)
+      end
+      deserializer.without_middlewares.deserialize_one(pgresult).tap do |persisted_event|
+        persisted_event.stream = stream
+      end
+    end
+
+    # @param stream [PgEventstore::Stream] persisted stream
+    # @return [void]
+    def update_stream_revision(stream, revision)
+      connection.with do |conn|
+        conn.exec_params(<<~SQL, [revision, stream.id])
+          UPDATE streams SET stream_revision = $1 WHERE id = $2
+        SQL
+      end
+    end
+
+    private
+
+    # @param stream [PgEventstore::Stream]
+    # @param options [Hash]
+    # @param offset [Integer]
+    # @return [PgEventstore::EventsFilteringQuery]
+    def events_filtering_builder(stream, options, offset: 0)
```

Comment:
Seems this method is too complex Are you using rubocop? I think it should be extracted to a builder class. The method name already implies that. The if inside the method alos implies it might make sense to create 2 different builder classes with a common base class. One for the all stream and one for specific streams.

Reply:
I wouldn't split that on two different classes, but would simply extract that to two separate methods. That can improve readability indeed. Will improve

Reply:
If you move to classes you can test nicely and get rid of the if.

Reply:
Check the implementation please. It is just a set of methods calls to compute the query. I extracted it into separated methods. You can recheck